src/services/identity/UsersService.ts

```diff
@@ -188,26 +210,98 @@ class UsersService {
   }
 
   async sendEmailOtp(email: string) {
-    const otp = this.otp.generate(email)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
+
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({ where: { email } })
+    if (!otpEntry) {
+      // create new entry
+      await this.createOtpEntry(email, OtpType.Email, hashedOtp)
+    } else {
+      await otpEntry?.update({
+        hashedOtp,
+        attempts: 0,
+        expiresAt: this.getOtpExpiry(),
+      })
+    }
 
     const subject = "One-Time Password (OTP) for IsomerCMS"
-    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${expiry} minutes. Please use this to verify your email address.</p>
+    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your email address.</p>
     <p>If your OTP does not work, please request for a new OTP.</p>
     <p>IsomerCMS Support Team</p>`
     await this.mailer.sendMail(email, subject, html)
   }
 
   async sendSmsOtp(mobileNumber: string) {
-    const otp = this.otp.generate(mobileNumber)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
+
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({
+      where: { mobileNumber },
+    })
+    if (!otpEntry) {
+      await this.createOtpEntry(mobileNumber, OtpType.Mobile, hashedOtp)
+    } else {
+      await otpEntry?.update({ hashedOtp, attempts: 0 })
+    }
 
-    const message = `Your OTP is ${otp}. It will expire in ${expiry} minutes. Please use this to verify your mobile number`
+    const message = `Your OTP is ${otp}. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your mobile number`
     await this.smsClient.sendSms(mobileNumber, message)
   }
 
-  verifyOtp(value: string, otp: string) {
-    return this.otp.verify(value, otp)
+  async verifyOtp(key: string, keyType: OtpType, otp: string) {
+    const otpFindParams =
+      keyType === OtpType.Email ? { email: key } : { mobileNumber: key }
+    const otpEntry = await this.otpRepository.findOne({ where: otpFindParams })
+
+    if (!otpEntry?.hashedOtp) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Hashed OTP not found")
+    }
+
+    if (otpEntry.attempts >= PARSED_MAX_NUM_OTP_ATTEMPTS) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Max number of attempts reached")
+    }
```

Comment:
I realised we're not verifying that the OTP has yet to expire - I think we should be checking against `otpEntry.expiresAt > new Date()`?

Reply:
good point, added!